packages/near-api-js/src/providers/wallet-rpc-provider.ts

```diff
@@ -0,0 +1,473 @@
+        super();
+        this._provider = provider;
+        this._provider.on('chainChanged', this.updateNetwork.bind(this));
+        this._provider.on('accountsChanged', this.updateAccount.bind(this));
+        this.init();
+    }
+
+    private async init() {
+        if(this._provider) {
+            const { chain_id } = await this.status();
+            this._network = chain_id;
+            this._account = await this._provider.getAccount();
+        }
+    }
+
+    /**
+     * update selected chain id
+     */
+    private updateNetwork(chainId: string) {
+        this._network = chainId;
+    }
+
+    /**
+     * update selected account
+     */
+    private updateAccount(account: { address: string; pubKeys: string[]}) {
+        this._account = account.address;
+        this._pubKeys = account.pubKeys;
+    }
+
+    /**
+     * Add a new listener (eg: chainChanged, accountsChanged)
+     */
+    on(message: string, listener: (...args: any[]) => void ): void {
+        if(this._provider) {
+            this._provider.on(message, listener);
+        }
+    }
+
+    /**
+     * Gets the RPC's status
+     * @see [https://docs.near.org/docs/develop/front-end/rpc#general-validator-status](https://docs.near.org/docs/develop/front-end/rpc#general-validator-status)
+     */
+    async status(): Promise<NodeStatusResult> {
+        return this.sendJsonRpc('status', []);
+    }
+
+    /**
+     * Sends a signed transaction to the RPC and waits until transaction is fully complete
+     * @see [https://docs.near.org/docs/develop/front-end/rpc#send-transaction-await](https://docs.near.org/docs/develop/front-end/rpc#general-validator-status)
+     *
+     * @param signedTransaction The signed transaction being sent
+     */
+    async sendTransaction(signedTransaction: SignedTransaction): Promise<FinalExecutionOutcome> {
+        const bytes = signedTransaction.encode();
+        return this.sendJsonRpc('broadcast_tx_commit', [Buffer.from(bytes).toString('base64')]);
+    }
+
+    /**
+     * Sends a signed transaction to the RPC and immediately returns transaction hash
+     * See [docs for more info](https://docs.near.org/docs/develop/front-end/rpc#send-transaction-async)
+     * @param signedTransaction The signed transaction being sent
+     * @returns {Promise<FinalExecutionOutcome>}
+     */
+    async sendTransactionAsync(signedTransaction: SignedTransaction): Promise<FinalExecutionOutcome> {
+        const bytes = signedTransaction.encode();
+        return this.sendJsonRpc('broadcast_tx_async', [Buffer.from(bytes).toString('base64')]);
+    }
+
+    /**
+     * Sign and Sends a transaction to the RPC and waits until transaction is fully complete
+     * @see {@link https://docs.near.org/docs/develop/front-end/rpc#send-transaction-await}
+     *
+     * @param transactions The transactions being sent
+     */
+    async signAndSendTransaction(transactions: Transaction[]): Promise<FinalExecutionOutcome> {
```

Comment:
What will end up calling this? Is `signAndSendTransation` a valid RPC method?

Reply:
The `signAndSendTransaction` method means signing and sending a transaction at once in a Wallet. If the wallet implemented this feature, user can implement `request` in WalletProvider and use it. The following is an example of using the WELLDONE Wallet.

```typescript
async request(args: RequestParams) {
  if (args.method === 'signAndSendTransaction') {
    args.method = 'dapp:sendTransaction';
  }
  const result = await (window as any).dapp.request('near', args);
  return result;
}
```
Please see the [docs](https://docs.welldonestudio.io/wallet/developer-guide/provider-api/method#sendTransaction) for an explanation of the `dapp:sendTransaction` method.